gse-util/src/main/java/com/powsybl/gse/util/GroovyCodeEditor.java

```diff
@@ -104,6 +118,48 @@ public GroovyCodeEditor(Scene scene) {
         codeArea.setOnSelectionDrag(p -> allowedDrag = true);
     }
 
+    private void setOnPaste(KeyEvent ke) {
+        final Clipboard clipboard = Clipboard.getSystemClipboard();
+        if (clipboard.hasString()) {
+            ke.consume();
+            String formatedClipBoardText = clipboard.getString().replaceAll(Character.toString('\t'), tabSpace(getTabSize()));
+            codeArea.insertText(codeArea.getCaretPosition(), formatedClipBoardText);
+        }
+    }
+
+    private void onTabPressed() {
+        codeArea.deletePreviousChar();
+        int currentLine = codeArea.getCaretSelectionBind().getParagraphIndex();
+        int fromlineStartToCaret = codeArea.getText(currentLine, 0, currentLine, codeArea.getCaretColumn()).length();
+        codeArea.insertText(codeArea.getCaretPosition(), tabSpace(tabSpacesToComplete(fromlineStartToCaret)));
+    }
+
+    public void setTabSize(int size) {
+        codeArea.tabSize = size;
+    }
+
+    public int getTabSize() {
+        return codeArea.tabSize;
+    }
+
+    private String tabSpace(int size) {
+        StringBuilder space = new StringBuilder();
+        for (int i = 0; i < size; i++) {
+            space.append(" ");
+        }
+        return space.toString();
+    }
+
+    private int tabSpacesToComplete(int length) {
```

Comment:
Should be static

This method seems complicated: what are you trying to do? To add the correct number of spaces from the current position to « complete » the tabulation?

int spacesToAdd = getTabSize() - (length % getTabSize());

Reply:
updated